AWS S3 Authentication and Fuzzing for Ceph RGW

Suyashd999 opened the pull request:
## Summary

This pull request adds support for AWS Signature Version 4 (SigV4) authentication in RESTler, enabling fuzzing of AWS S3-compatible endpoints that require AWS request signing.

## Details

- Introduced a sample authentication module `aws_sigv4_auth.py` (originally located in `restler/utils`) to demonstrate how RESTler can sign requests using AWS S3 credentials (`AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`), leveraging `boto3` and `botocore`.
- After review feedback, the sample file has been removed from `restler/utils` to avoid including service-specific logic in the RESTler core codebase.
- The full sample, with usage and dependency details, is now tracked in a separate GitHub issue for discoverability and ongoing maintenance.

## Usage

- To enable request signing for AWS S3 or other SigV4-compatible endpoints, follow the [`aws_sigv4_auth.py` example in Issue #987](https://github.com/microsoft/restler-fuzzer/issues/987).
- **Install dependencies:** `pip install boto3 botocore`
- Put your AWS credentials in the appropriate environment variables or pass them to the script.
- Integrate the sample into your test deployment or adapt as needed for your authentication flow.

## Related

- Issue: [#987 – AWS SigV4 Authentication Example for RESTler](https://github.com/microsoft/restler-fuzzer/issues/987)

Suyashd999 committed:
Current configuration for testing RGW with Restler

yuvalif commented on restler/utils/aws_sigv4_auth.py:
```diff
@@ -0,0 +1,148 @@
+# Set up logging file location
+FILE_PATH = "/home/suyash/Downloads/boto.log"
+
+logging.basicConfig(
```
please use the existing restler logging mechanism

yuvalif commented:
@Suyashd999 
* can you please rename the PR so its name is not RGW/Ceph specific
* please squash commits

yuvalif commented:
@marina-p could please have another look at the PR?

marina-p commented on restler/engine/transport_layer/messaging.py:
```diff
@@ -208,6 +208,35 @@ def _append_to_header(message, content):
+                auth_module = Settings().authentication['module']
+                signing_function = auth_module.get('function', 'sign_request')
+                signing_module = __import__(auth_module['name'], fromlist=[signing_function])
+                sign_request = getattr(signing_module, signing_function)
```
Could you please use the helper `import_attr` in import_utilities.py?

marina-p commented on restler/utils/aws_sigv4_auth.py:
```diff
@@ -0,0 +1,124 @@
+import boto3
+from botocore.auth import SigV4Auth
+from botocore.awsrequest import AWSRequest
+from botocore.credentials import Credentials
```
This module needs to be part of your own team's deployment - RESTler does not maintain authentication for specific services in the code base.

yuvalif replied:
maybe this could be added under an "examples" directory, to give users an idea of how to configure a signer?

marina-p replied:
Ideally such samples would be working samples with their own set of dependencies that could be installed, tested etc.  Since we don't have such samples today, could you please open an issue and attach this there to track adding such examples, but keep it separate from this PR so as not to block the PR?

marina-p reviewed (changes requested):
:clock1:

marina-p commented on restler/engine/transport_layer/messaging.py:
```diff
@@ -208,6 +208,35 @@ def _append_to_header(message, content):
 
+        # Add request signing if enabled in authentication settings    
+        if Settings().authentication and Settings().authentication.get('module', {}).get('signing'):
+            try:                
```
Sorry to be catching this at PR time - could you please make this consistent with the other authentication dictionary structure, so ```"authentication": {  "signing": {"module": {"file"/"function"/...}}}```.  See SettingsFile.md for the structure.   #Resolved

marina-p commented on restler/engine/transport_layer/messaging.py:
```diff
@@ -208,6 +208,35 @@ def _append_to_header(message, content):
                 message = _append_to_header(message, f"x-restler-sequence-id: {sequence_id}")
 
+        # Add request signing if enabled in authentication settings    
+        if Settings().authentication and Settings().authentication.get('module', {}).get('signing'):
```
>signing [](http://example.com/codeflow?start=90&length=7)

Please update the docs - Authentication.md (quick 1-2 sentences description is sufficient), and SettingsFile.md (sample settings object that works when copied into the engine settings with updated paths). #Resolved

marina-p commented:
Ideally, some tests should be added for this using the unit test server.  This can be done in a future PR if you've already validated these changes - could you please open an issue to track this?

---
In reply to: [3395031310](https://github.com/microsoft/restler-fuzzer/pull/962#issuecomment-3395031310) [](http://example.com/codeflow?ancestors=3395031310)

yuvalif commented:
> Ideally, some tests should be added for this using the unit test server. This can be done in a future PR if you've already validated these changes - could you please open an issue to track this?
> 
> In reply to: [3395031310](https://github.com/microsoft/restler-fuzzer/pull/962#issuecomment-3395031310)

this configuration and signing mechanism is tested against our server.

if we want to add a dedicated test to restler we probably dont want to test against a full S3 gateway.
several options:
* just implement somnething simple: e.g. hash the header and add the hash as a filed on the client, the server will look for the hash field and verify it is correct
* use the IETF standard for HTTP signing. there is a pyhon library that already implements that: https://pypi.org/project/http-message-signatures/ . note that it has the verifier code built into the lib
* implement a dummy S3 python server that just verify the signature

Suyashd999 committed:
feat: Add request signing authentication support with consistent configuration structure

- Import and use import_attr helper from utils.import_utilities for cleaner module loading
- Update AWS SigV4 authentication module comments to reflect new config structure
- Document request signing authentication in Authentication.md and SettingsFile.md

Suyashd999 commented:
hey @marina-p, can you review the PR once again. Thanks!

marina-p commented:
Done - just one remaining comment RE: restler/utls, everything else looks good!

---
In reply to: [3408292266](https://github.com/microsoft/restler-fuzzer/pull/962#issuecomment-3408292266) [](http://example.com/codeflow?ancestors=3408292266)

Suyashd999 commented:
> Done - just one remaining comment RE: restler/utls, everything else looks good!
> 
> In reply to: [3408292266](https://github.com/microsoft/restler-fuzzer/pull/962#issuecomment-3408292266)

Hi @marina-p! I have made that change as well. Please do check once ! Thankyou !!

Suyashd999 commented:
> Hi @marina-p! I have made that change as well. Please do check once ! Thankyou !!

Hey @marina-p, a reminder here. Thankyou!

marina-p reviewed (approved):
:shipit:

yuvalif commented:
> :shipit:

@marina-p anything else that needs to be done so before this is merged?

marina-p commented:
Hi @yuvalif, the code looks good.  The contributor license agreement check needs to pass for it to be merged. @Suyashd999, is it possible for you to sign per the required steps?

marina-p commented:
See the 'Contributing' section in the README for instructions.

Suyashd999 commented:
@microsoft-github-policy-service agree

yuvalif commented:
@marina-p any update on this PR?

adamemerson commented:
@Suyashd999 Hi, could you rebase this, please?

yuvalif commented:
> @Suyashd999 Hi, could you rebase this, please?

dont think there are any merge conflicts. this is mainly a matter of running/passing the checks
